src/engine.cc

```diff
@@ -210,20 +220,57 @@ void EngineController::SetPosition(const std::string& fen,
 }
 
 void EngineController::Go(const GoParams& params) {
-  if (!tree_) {
-    SetPosition(ChessBoard::kStartingFen, {});
+  go_params_ = params;
+
+  ThinkingInfo::Callback info_callback(info_callback_);
+
+  if(current_position_) {
+    if (params.ponder && !(*current_position_).moves.empty()) {
+      std::vector<std::string> moves((*current_position_).moves);
+      std::string ponder_move = moves.back();
+      moves.pop_back();
+      SetupPosition((*current_position_).fen, moves);
```

Comment:
The problem with that approach is that first when position is set up, useful (for tree reuse) subtrees are chopped.
And then when `go ponder` comes it, it turns out that those subtrees were useful, and how the entire node is useless too, so it's trimmed.
Shortly speaking, ponder kills tree reuse.
I'm trying to come up with elegant-ish way to solve that, ideas are welcome. We can move all that deallocation to `go` command, but then why are those `isready` there at all..

Reply:
Where exactly are useful subtrees chopped? The `SetupPosition` should set the position currently  on the board (the uci ponder position minus the last move). That's the whole point of the deferred position setup.
The unmodified `SetPosition` function would indeed have chopped useful subtrees.

Reply:
Yes, I know. That's why I made sure that the new position is never shorter than the old one. (At least I think I did, and the node counts when testing this PR suggested that indeed the tree was correctly re-used.)

Reply:
Yeah, I thought that you reset tree in setposition too, sorry for the confusion!